### PR TITLE
lxd: Add project support

### PR DESCRIPTION
**Pull request by stgraber**
This is an initial attempt at adding some level of LXD project support in Juju.

### CODE REVIEW

**Comment by stgraber** [comment]
For this one, I'll need some help because the change doesn't actually work.

I can set the new config key and I see it in my local `juju show-cloud` output, but when adding it to a controller, the config key disappears and so Juju deploys into the default project instead. My controller is running the development code so I'm quite confused as to what's going on.

I suspect it may be missing plumbing in the LXD provider as it never had a config key before, but after a couple of hours of going through what felt like hundreds of layers of indirection, I came up empty ;)

**Comment by stgraber** [comment]
and I fully realize the current code is hackish and won't pass the testsuite, that's fine, I just want to see it work before I go and fix all the other code paths to match :)

File: container/lxd/connection.go [hunk 1]
```diff
@@ -141,7 +143,14 @@ func ConnectRemote(spec ServerSpec) (lxd.ContainerServer, error) {
 		return nil, errors.Trace(err)
 	}
 	client, err := lxd.ConnectLXD(uri, spec.connectionArgs)
-	return client, errors.Trace(err)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if spec.Project != "" {
+		client.UseProject(spec.Project)
```
Comment (manadart) [inline container/lxd/connection.go:150]: I got this to work with:
```go
client = client.UseProject(spec.Project)
```

Comment (stgraber) [reply container/lxd/connection.go:150]: Doh!
That would explain that part :)

I'm still confused as to why I couldn't get the new config key to stick when adding a new cloud to a shared controller though. But maybe it's my testing that was wrong somehow.

Comment (manadart) [reply container/lxd/connection.go:150]: I did it with a fresh bootstrap, using `--config`, which adds it to model defaults.
`juju model-config` showed me the project I configured.
If you can give more detail on what you did I can try to chase it down...

Comment (stgraber) [reply container/lxd/connection.go:150]: I had an existing juju controller which I upgraded to my local dev version through `juju upgrade-controller --build-agent`, I then created a local cloud YAML for my remote LXD cluster containing `project: foo` in the config section, then added that and matching credentials locally first and then to the existing controller.

This did show up as a new cloud on the controller but the `project` config key did not make it over.

(Sorry, that's from memory, I don't have that juju dev environment here anymore)

**Comment by manadart** [comment]
Thanks for putting this up, and sorry it has taken so long.
I cherry-picked your commit onto #13111 then changed the implementation somewhat.
We can iterate from there.

**Pull request closed without merging**